client/hostvolumemanager/host_volume_plugin.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package hostvolumemanager
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/hashicorp/go-hclog"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+)
+
+type HostVolumePlugin interface {
+	Version(ctx context.Context) (string, error)
+	Create(ctx context.Context, req *cstructs.ClientHostVolumeCreateRequest) (*hostVolumePluginCreateResponse, error)
+	Delete(ctx context.Context, req *cstructs.ClientHostVolumeDeleteRequest) error
+	// TODO(db): update? resize? ??
+}
+
+type hostVolumePluginCreateResponse struct {
+	Path      string            `json:"path"`
+	SizeBytes int64             `json:"bytes"`
+	Context   map[string]string `json:"context"` // metadata
+}
+
+var _ HostVolumePlugin = &HostVolumePluginMkdir{}
+
+type HostVolumePluginMkdir struct {
+	ID         string
+	TargetPath string
+
+	log hclog.Logger
+}
+
+func (p *HostVolumePluginMkdir) Version(_ context.Context) (string, error) {
+	return "0.0.1", nil
+}
+
+func (p *HostVolumePluginMkdir) Create(_ context.Context,
+	req *cstructs.ClientHostVolumeCreateRequest) (*hostVolumePluginCreateResponse, error) {
+
+	path := filepath.Join(p.TargetPath, req.ID)
+	p.log.Debug("CREATE: default host volume plugin", "target_path", path)
+
+	err := os.Mkdir(path, 0o700)
+	if err != nil {
+		return nil, err
+	}
+
+	return &hostVolumePluginCreateResponse{
+		Path:      path,
+		SizeBytes: 0,
+		Context:   map[string]string{},
+	}, nil
+}
+
+func (p *HostVolumePluginMkdir) Delete(_ context.Context, req *cstructs.ClientHostVolumeDeleteRequest) error {
+	path := filepath.Join(p.TargetPath, req.ID)
+	p.log.Debug("DELETE: default host volume plugin", "target_path", path)
+	return os.RemoveAll(path)
+}
+
+var _ HostVolumePlugin = &HostVolumePluginExternal{}
+
+type HostVolumePluginExternal struct {
+	ID         string
+	Executable string
+	TargetPath string
+
+	log hclog.Logger
+}
+
+func (p *HostVolumePluginExternal) Version(_ context.Context) (string, error) {
+	return "0.0.1", nil // TODO(db): call the plugin, use in fingerprint
+}
+
+func (p *HostVolumePluginExternal) Create(ctx context.Context,
+	req *cstructs.ClientHostVolumeCreateRequest) (*hostVolumePluginCreateResponse, error) {
+
+	params, err := json.Marshal(req.Parameters) // TODO(db): if this is nil, then PARAMETERS env will be "null"
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling volume pramaters: %w", err)
+	}
```

Comment:
I went back and forth on this. a script with `set -u` will explode if it's not set, making it more difficult to gracefully handle the `null` case.  another option would be to send an empty map `{}` but that feels less explicit. :shrug:

Reply:
Shouldn't it explode though? I feel like a given plugin script is likely to either (a) always want parameters, or (b) never want parameters. In scenario (a), a script is expecting a value to always be present and it's not set by the volume spec, which is an error. In scenario (b), the script isn't going to read the parameters anyways.

Reply:
an example of an optional Parameter I can think of: `chmod` to some user, if provided, otherwise leave it as the Nomad user creating it (`root`, probably).  or optional extra mount options.  I imagine there may be others.